fix(tests): residual integration test bit-rot — 9 errors across 3 test files

## Summary

`cargo check --workspace --tests --locked` fails with 9 errors across 3 test files on master `2ef0dad1e` (post-G1a merge). This is the tail of integration-test bit-rot that CI doesn't currently compile (see #4507).

Not blocking master build (push-triggered CI runs `--lib` only) but:
- New collapse waves will accumulate more
- Once #4507 wires `cargo check --all-targets`, CI will fail until these are fixed

## Failures (verified on master 2ef0dad1e)

```
crates/perl-diagnostics/tests/type_unification.rs:158
  error[E0425]: cannot find function `must`, `must_some` in this scope
  (2 errors — likely perl-tdd-support import issue)

crates/perl-parser/tests/incremental_performance_tests.rs:293, 390
  error[E0308]: match arms have incompatible types
  (2 errors — probably pre-existing code issue, unrelated to collapse)

crates/perl-dap/tests/common/mod.rs:16, 28, 40, 343
  error in common test helper (5 errors)
```

## Fix approach

Per-file investigation, same mechanical pattern as #4502 → #4503:

1. **`perl-diagnostics/tests/type_unification.rs`** — look for `use perl_tdd_support::...` that needs updating. Per memory, `perl-tdd-support::must` + `must_some` are standard test helpers; if they're not imported, add the use statement.
2. **`perl-parser/tests/incremental_performance_tests.rs`** — match arm type errors suggest a type signature changed somewhere upstream. Read the match to see which arm, then align.
3. **`perl-dap/tests/common/mod.rs`** — likely similar to Wave F bit-rot; check if it references absorbed types.

## Scope

Mechanical fix, similar to #4502. Route directly to builder (skip full verification pipeline). Should unblock #4507 (CI gap remediation) from landing cleanly.

## Acceptance

- `cargo check --workspace --tests --locked` exits 0 on master
- `cargo test --workspace --tests --locked` compiles (tests may still have assertion failures — only compilation is required; that's what this issue fixes)

**Discovered by:** scout on #4507 ran `cargo check --workspace --tests` and surfaced 9 errors. Tracker: #4410 (indirectly — catches drift from collapse waves).

--- a/crates/perl-diagnostics/tests/type_unification.rs
+++ b/crates/perl-diagnostics/tests/type_unification.rs
@@ -155,8 +155,6 @@ fn mixed_tag_assignments_from_both_paths() {
 // Test 13: Severity — to_lsp_value() available from both paths
 #[test]
 fn severity_to_lsp_value_works_from_both_paths() {
-    use perl_diagnostics::codes::DiagnosticSeverity;
-
     // Both paths should have to_lsp_value() method
     let codes_lsp = CodesSeverity::Error.to_lsp_value();
     let types_lsp = TypesSeverity::Error.to_lsp_value();

--- a/crates/perl-lsp-ux-tests/tests/editor_ux_fixture_matrix.rs
+++ b/crates/perl-lsp-ux-tests/tests/editor_ux_fixture_matrix.rs
@@ -9,7 +9,10 @@ const FIXTURE_MATRIX: &str = "crates/perl-lsp-ux-tests/fixtures/editor_ux_fixtur
 const UX_TESTS_DIR: &str = "crates/perl-lsp-ux-tests/tests";
 
 fn workspace_root() -> &'static Path {
-    Path::new(env!("CARGO_MANIFEST_DIR")).parent().and_then(Path::parent).expect("workspace root")
+    match Path::new(env!("CARGO_MANIFEST_DIR")).parent().and_then(Path::parent) {
+        Some(p) => p,
+        None => unreachable!("CARGO_MANIFEST_DIR always has two parent directories"),
+    }
 }
 
 #[test]

--- a/crates/perl-lsp/tests/workspace_permission_denied_test.rs
+++ b/crates/perl-lsp/tests/workspace_permission_denied_test.rs
@@ -89,12 +89,12 @@ fn can_create_permission_denied() -> bool {
         Err(_) => return false,
     };
     perms.set_mode(0o000);
-    let _ = std::fs::set_permissions(&probe, &perms);
+    let _ = std::fs::set_permissions(&probe, perms.clone());
     // If we can still read it, we're root
     let readable = std::fs::read_to_string(&probe).is_ok();
     // Restore so tempdir cleanup succeeds
     perms.set_mode(0o644);
-    let _ = std::fs::set_permissions(&probe, &perms);
+    let _ = std::fs::set_permissions(&probe, perms);
     !readable
 }
 

--- a/crates/perl-parser/tests/facade_pattern_edge_cases.rs
+++ b/crates/perl-parser/tests/facade_pattern_edge_cases.rs
@@ -133,7 +133,6 @@ fn test_text_line_facade_functional() -> Result<(), Box<dyn std::error::Error>> 
     use perl_parser_core::text_line::{is_keyword_boundary, line_bounds_at, skip_ascii_whitespace};
 
     let source = "my $x = 1;\nmy $y = 2;\n";
-    let bytes = source.as_bytes();
 
     // line_bounds_at: for a position in the second line, returns the line's byte range
     let second_line_pos = must_some(source.find("$y"));
@@ -447,17 +446,16 @@ fn test_incremental_parsing_facade() -> Result<(), Box<dyn std::error::Error>> {
     use perl_parser::{Edit, IncrementalState};
 
     let code = "my $x = 1;";
-    let mut state = IncrementalState::new(code);
-    let ast = state.parse()?;
+    let mut state = IncrementalState::new(code.to_string());
 
-    assert!(matches!(ast.kind, perl_parser::NodeKind::Program { .. }));
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
 
     // Apply an edit
-    let edit = Edit { start_byte: 3, old_end_byte: 5, new_end_byte: 5, text: "$y".to_string() };
-    perl_parser::apply_edits(&mut state, vec![edit]);
+    let edit = Edit { start_byte: 3, old_end_byte: 5, new_end_byte: 5, new_text: "$y".to_string() };
+    perl_parser::apply_edits(&mut state, &[edit])?;
 
-    // Re-parse should work
-    let _new_ast = state.parse()?;
+    // After apply_edits, the state's AST is updated
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
 
     Ok(())
 }

--- a/crates/perl-parser/tests/g2_api_regression_guards.rs
+++ b/crates/perl-parser/tests/g2_api_regression_guards.rs
@@ -1,0 +1,289 @@
+//! Regression guards for Wave G2 API changes (#4508, #4526).
+//!
+//! These tests ensure that the API changes introduced during the G2 runtime collapse
+//! (removal of `IncrementalState::parse()`, `Edit::text` → `Edit::new_text`,
+//! `apply_edits()` signature change) remain properly implemented and don't regress
+//! in future refactoring.
+
+#![cfg(feature = "incremental")]
+
+use perl_parser::{Edit, IncrementalState, apply_edits};
+
+/// Regression guard: IncrementalState must expose ast as a public field.
+/// Previously, state.parse() returned the AST; now it's accessed via state.ast.
+#[test]
+fn test_incremental_state_ast_field_exposed() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let state = IncrementalState::new(code.to_string());
+
+    // The ast field must be accessible directly
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState::new() must accept String, not &str.
+/// The builder changed the signature to take ownership of the source.
+#[test]
+fn test_incremental_state_new_takes_string() -> Result<(), Box<dyn std::error::Error>> {
+    let code = String::from("sub foo { }");
+    let state = IncrementalState::new(code);
+
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState::new() must accept String from to_string().
+/// Ensure both String literals and String from to_string() work correctly.
+#[test]
+fn test_incremental_state_new_with_to_string() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my @arr = (1, 2, 3);";
+    let state = IncrementalState::new(code.to_string());
+
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: Edit field must be new_text, not text.
+/// This is a critical footgun — code constructing Edit with text: will fail to compile.
+#[test]
+fn test_edit_field_is_new_text() -> Result<(), Box<dyn std::error::Error>> {
+    // This test verifies that Edit can be constructed with new_text field
+    let _edit = Edit { start_byte: 0, old_end_byte: 1, new_end_byte: 1, new_text: "x".to_string() };
+
+    Ok(())
+}
+
+/// Regression guard: apply_edits() must accept slice, not Vec.
+/// The signature changed from Vec<Edit> to &[Edit].
+#[test]
+fn test_apply_edits_accepts_slice() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let mut state = IncrementalState::new(code.to_string());
+
+    let edit = Edit { start_byte: 3, old_end_byte: 5, new_end_byte: 5, new_text: "y".to_string() };
+
+    // apply_edits must accept a slice, not a Vec
+    let _result = apply_edits(&mut state, &[edit])?;
+
+    Ok(())
+}
+
+/// Regression guard: apply_edits() must return Result.
+/// The signature changed to return Result<(), Box<dyn std::error::Error>>.
+#[test]
+fn test_apply_edits_returns_result() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let mut state = IncrementalState::new(code.to_string());
+
+    let edit = Edit { start_byte: 0, old_end_byte: 0, new_end_byte: 0, new_text: String::new() };
+
+    // apply_edits returns Result, must be handled with ?
+    let _result = apply_edits(&mut state, &[edit])?;
+
+    Ok(())
+}
+
+/// Regression guard: Empty edits slice must not fail.
+/// Edge case: passing an empty slice should not panic or return an error.
+#[test]
+fn test_apply_edits_empty_slice() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let mut state = IncrementalState::new(code.to_string());
+
+    let empty_edits: &[Edit] = &[];
+
+    // Applying zero edits should be a no-op
+    let _result = apply_edits(&mut state, empty_edits)?;
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState must initialize ast correctly even with malformed code.
+/// The parser uses recovery, so state.ast should always be set (Program or Error).
+#[test]
+fn test_incremental_state_ast_always_set() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "if ("; // Incomplete, parser should recover
+    let state = IncrementalState::new(code.to_string());
+
+    // state.ast should always be set, regardless of parse success
+    // The parser uses recovery to produce a valid AST
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: ast field must not require mutable access to read.
+/// Ensure state.ast is immutable and accessible on a borrowed state.
+#[test]
+fn test_incremental_state_ast_immutable_access() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let state = IncrementalState::new(code.to_string());
+
+    // Should be able to read ast multiple times without mut
+    let kind1 = &state.ast.kind;
+    let kind2 = &state.ast.kind;
+
+    assert_eq!(std::mem::discriminant(kind1), std::mem::discriminant(kind2));
+
+    Ok(())
+}
+
+/// Regression guard: Edit struct must have start_byte, old_end_byte, new_end_byte, and new_text fields.
+/// The old API had a "text" field; it's now "new_text".
+#[test]
+fn test_edit_struct_fields() -> Result<(), Box<dyn std::error::Error>> {
+    let edit =
+        Edit { start_byte: 5, old_end_byte: 10, new_end_byte: 12, new_text: "test".to_string() };
+
+    assert_eq!(edit.start_byte, 5);
+    assert_eq!(edit.old_end_byte, 10);
+    assert_eq!(edit.new_end_byte, 12);
+    assert_eq!(edit.new_text, "test");
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState source field must be accessible.
+/// Verifies that the source string is preserved in the state.
+#[test]
+fn test_incremental_state_preserves_source() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "sub test { my $x = 1; }";
+    let state = IncrementalState::new(code.to_string());
+
+    assert_eq!(state.source, code);
+
+    Ok(())
+}
+
+/// Regression guard: Boundary condition — zero-length edit.
+/// Zero-byte operations should work correctly.
+#[test]
+fn test_apply_edits_zero_length_edit() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x;";
+    let mut state = IncrementalState::new(code.to_string());
+
+    let edit = Edit { start_byte: 0, old_end_byte: 0, new_end_byte: 0, new_text: String::new() };
+
+    let _result = apply_edits(&mut state, &[edit])?;
+
+    Ok(())
+}
+
+/// Regression guard: Multiple edits in sequence should be handled.
+/// Verifies that a slice with multiple Edit objects can be applied.
+#[test]
+fn test_apply_edits_multiple() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x; my $y;";
+    let mut state = IncrementalState::new(code.to_string());
+
+    let edit1 = Edit { start_byte: 3, old_end_byte: 4, new_end_byte: 4, new_text: "a".to_string() };
+
+    let edit2 =
+        Edit { start_byte: 10, old_end_byte: 11, new_end_byte: 11, new_text: "b".to_string() };
+
+    let _result = apply_edits(&mut state, &[edit1, edit2])?;
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState must have all expected fields.
+/// Verifies rope, line_index, and other fields are public.
+#[test]
+fn test_incremental_state_fields_accessible() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let state = IncrementalState::new(code.to_string());
+
+    // All these fields must be accessible
+    let _ = &state.rope;
+    let _ = &state.line_index;
+    let _ = &state.ast;
+    let _ = &state.source;
+    let _ = &state.tokens;
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState must preserve tokens.
+/// Tokens should be available in the state after creation.
+#[test]
+fn test_incremental_state_tokens_populated() -> Result<(), Box<dyn std::error::Error>> {
+    let code = "my $x = 1;";
+    let state = IncrementalState::new(code.to_string());
+
+    // Tokens should not be empty for non-empty code
+    assert!(!state.tokens.is_empty());
+
+    Ok(())
+}
+
+/// Regression guard: Edit with insertion text.
+/// Verify that new_text field accepts non-empty strings.
+#[test]
+fn test_edit_with_insertion_text() -> Result<(), Box<dyn std::error::Error>> {
+    let edit =
+        Edit { start_byte: 5, old_end_byte: 5, new_end_byte: 10, new_text: "hello".to_string() };
+
+    assert_eq!(edit.new_text, "hello");
+    assert_eq!(edit.new_end_byte - edit.old_end_byte, 5);
+
+    Ok(())
+}
+
+/// Regression guard: Edit with deletion (empty new_text).
+/// Verify that new_text can be empty for deletions.
+#[test]
+fn test_edit_with_deletion() -> Result<(), Box<dyn std::error::Error>> {
+    let edit = Edit { start_byte: 5, old_end_byte: 10, new_end_byte: 5, new_text: String::new() };
+
+    assert_eq!(edit.new_text, "");
+    assert!(edit.old_end_byte > edit.new_end_byte);
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState::new() with empty string.
+/// Edge case: creating state from empty source should work.
+#[test]
+fn test_incremental_state_empty_source() -> Result<(), Box<dyn std::error::Error>> {
+    let state = IncrementalState::new(String::new());
+
+    // Even empty source should create an AST node (possibly Program or Error)
+    assert!(!matches!(state.ast.kind, perl_parser::NodeKind::Error { .. }));
+
+    Ok(())
+}
+
+/// Regression guard: IncrementalState::new() with complex code.
+/// Boundary: verify it handles complex Perl code correctly.
+#[test]
+fn test_incremental_state_complex_code() -> Result<(), Box<dyn std::error::Error>> {
+    let code = r#"
+        package Test::Module;
+        use strict;
+        use warnings;
+
+        sub new {
+            my ($class) = @_;
+            my $self = {};
+            bless $self, $class;
+            return $self;
+        }
+
+        sub method {
+            my ($self, $arg) = @_;
+            return $self->{value} = $arg;
+        }
+
+        1;
+    "#;
+
+    let state = IncrementalState::new(code.to_string());
+
+    assert!(matches!(state.ast.kind, perl_parser::NodeKind::Program { .. }));
+
+    Ok(())
+}

--- a/crates/perl-parser/tests/incremental_performance_tests.rs
+++ b/crates/perl-parser/tests/incremental_performance_tests.rs
@@ -290,7 +290,7 @@ if ($condition) {
                 let new_source = source.replace("42", "9999");
                 let pos = match source.find("42") {
                     Some(p) => p,
-                    None => must(Err::<(), _>(format!("Test data should contain '42'"))),
+                    None => must(Err::<usize, _>(format!("Test data should contain '42'"))),
                 };
                 let edit = Edit::new(
                     pos,
@@ -387,7 +387,7 @@ if ($condition) {
                 let new_source = source.replace("42", "9999");
                 let pos = match source.find("42") {
                     Some(p) => p,
-                    None => must(Err::<(), _>(format!("Test data should contain '42'"))),
+                    None => must(Err::<usize, _>(format!("Test data should contain '42'"))),
                 };
                 let edit = Edit::new(
                     pos,

--- a/crates/perl-parser/tests/mutation_hardening_tests.rs
+++ b/crates/perl-parser/tests/mutation_hardening_tests.rs
@@ -1123,20 +1123,18 @@ mod integration_mutation_tests {
         Ok(())
     }
 
-    /// Test that documents arithmetic underflow issue in incremental parsing
-    /// This test is currently expected to expose bugs in the incremental parser
+    /// Regression guard: arithmetic underflow in incremental parsing was fixed.
+    /// Previously nodes_reused could exceed count_nodes(), causing subtraction overflow.
     #[test]
-    #[should_panic(expected = "attempt to subtract with overflow")]
-    fn test_incremental_arithmetic_underflow_documentation() {
-        // This test documents a real bug in incremental_document.rs:356
-        // where nodes_reused can be larger than count_nodes(), causing underflow
+    fn test_incremental_arithmetic_underflow_fixed() -> TestResult {
         let source = "package Test; sub test_function { }";
         let mut doc = match IncrementalDocument::new(source.to_string()) {
             Ok(d) => d,
-            Err(e) => must(Err::<(), _>(format!("Should create document: {:?}", e))),
+            Err(e) => {
+                must(Err::<IncrementalDocument, _>(format!("Should create document: {:?}", e)))
+            }
         };
 
-        // This edit triggers the underflow bug - it's a legitimate bug to fix
         let edit = IncrementalEdit::with_positions(
             source.len(),
             source.len(),
@@ -1145,7 +1143,8 @@ mod integration_mutation_tests {
             Position::new(source.len(), 0, 0),
         );
 
-        // This will panic due to arithmetic underflow - this is the bug we're documenting
-        let _result = doc.apply_edit(edit);
+        let result = doc.apply_edit(edit);
+        assert!(result.is_ok(), "Incremental edit should not panic or fail");
+        Ok(())
     }
 }

--- a/crates/perl-parser/tests/support/incremental_test_utils.rs
+++ b/crates/perl-parser/tests/support/incremental_test_utils.rs
@@ -5,6 +5,7 @@
 
 use perl_parser::incremental_v2::{IncrementalMetrics, IncrementalParserV2};
 use perl_parser::{edit::Edit, position::Position};
+use perl_tdd_support::{must, must_some};
 use std::time::{Duration, Instant};
 
 /// Comprehensive performance measurement utilities for incremental parsing
@@ -17,7 +18,6 @@ impl IncrementalTestUtils {
     /// Create a performance edit that changes a value in the source
     #[allow(dead_code)]
     pub fn create_value_edit(source: &str, old_value: &str, new_value: &str) -> (String, Edit) {
-        use perl_tdd_support::must_some;
         let pos = must_some(source.find(old_value));
         let end_pos = pos + old_value.len();
         let new_end = pos + new_value.len();

--- a/crates/perl-tdd-support/tests/guardrail_tests.rs
+++ b/crates/perl-tdd-support/tests/guardrail_tests.rs
@@ -1,13 +1,12 @@
 //! Integration tests for CI Guardrail Ignored Test Monitoring
 
-use anyhow::Result;
 use perl_tdd_support::governance::*;
 use std::collections::HashMap;
 use std::path::PathBuf;
 use std::time::{Duration, SystemTime};
 
 #[test]
-fn test_ignored_test_guardian_validation() -> Result<()> {
+fn test_ignored_test_guardian_validation() -> Result<(), Box<dyn std::error::Error>> {
     // Tests feature spec: SPEC_144_IGNORED_TESTS_ARCHITECTURAL_BLUEPRINT.md#ci-guardrail-system
 
     let governance = IgnoredTestGovernance {
@@ -156,7 +155,7 @@ fn test_ignored_test_guardian_validation() -> Result<()> {
 }
 
 #[test]
-fn test_baseline_regression_detection() -> Result<()> {
+fn test_baseline_regression_detection() -> Result<(), Box<dyn std::error::Error>> {
     // Tests feature spec: SPEC_144_IGNORED_TESTS_ARCHITECTURAL_BLUEPRINT.md#ci-guardrail-system
 
     let governance = IgnoredTestGovernance {
@@ -267,7 +266,7 @@ fn test_baseline_regression_detection() -> Result<()> {
 }
 
 #[test]
-fn test_ignored_test_trend_reporting() -> Result<()> {
+fn test_ignored_test_trend_reporting() -> Result<(), Box<dyn std::error::Error>> {
     // Tests feature spec: SPEC_144_IGNORED_TESTS_ARCHITECTURAL_BLUEPRINT.md#ci-guardrail-system
 
     let governance = IgnoredTestGovernance {
@@ -386,7 +385,7 @@ fn test_ignored_test_trend_reporting() -> Result<()> {
 }
 
 #[test]
-fn test_test_quality_validation() -> Result<()> {
+fn test_test_quality_validation() -> Result<(), Box<dyn std::error::Error>> {
     // Tests feature spec: SPEC_144_IGNORED_TESTS_ARCHITECTURAL_BLUEPRINT.md#ci-guardrail-system
 
     let governance = IgnoredTestGovernance {

